node_modules/oae-principals/lib/api.user.js

```diff
@@ -119,63 +126,275 @@ var createUser = module.exports.createUser = function(ctx, tenantAlias, displayN
+ * @param  {Object}         callback.err                An error that occurred, if any
+ * @param  {Object}         callback.principalIds       Object mapping each principal of `principals` to a `principalId`
+ */
+var getOrCreateGuestUsers = module.exports.getOrCreateGuestUsers = function(ctx, principals, callback) {
+    var principalIdsOrEmails = _.partition(principals, function(principal) {
+        return (AuthzUtil.isUserId(principal) || AuthzUtil.isGroupId(principal));
+    });
+    var emails = principalIdsOrEmails[1];
+
+    // Validate the non-principal ids are actually emails
+    var validator = new Validator();
+    _.each(emails,function(email) {
+        validator.check(email, {'code': 400, 'msg': 'The specified principal is not a valid email addres'}).isEmail();
+    });
+    if (validator.hasErrors()) {
+        return callback(validator.getFirstError());
     }
 
-    // Create the user
-    Cassandra.runQuery(q.query, q.parameters, function (err) {
+    // Create the guest user accounts
+    return _createGuestUsers(ctx, principals, callback);
+};
+
+/**
+ * Given a set of strings, create (or retrieve) a guest account for the strings that
+ * are email addresses. If the string is not an email address, it's expected to be
+ * a user or group identifier and returned as is.
+ *
+ * @param  {Context}        ctx                         Standard context object containing the current user and the current tenant
+ * @param  {String[]}       principals                  An array of principal ids to retrieve mixed with email address for which guest accounts should be created
+ * @param  {Function}       callback                    Standard callback function
+ * @param  {Object}         callback.err                An error that occurred, if any
+ * @param  {Object}         callback.principalIds       Object mapping each principal of `principals` to a `principalId`
+ * @api private
+ */
+var _createGuestUsers = function(ctx, principals, callback, _principalIds) {
+    _principalIds = _principalIds || {};
+    if (principals.length === 0) {
+        return callback(null, _principalIds);
+    }
+
+    var principal = principals.pop();
+    // If the `principal` is a user or group id, we can move on to the next one
+    if (AuthzUtil.isUserId(principal) || AuthzUtil.isGroupId(principal)) {
+        _principalIds[principal] = principal;
+        return _createGuestUsers(ctx, principals, callback, _principalIds);
+    }
+
+    // The `principal` is an email
+    createGuestUser(ctx, principal, function(err, user) {
         if (err) {
             return callback(err);
         }
 
-        var createdUser = new PrincipalsModel.User(tenantAlias, id, displayName, opts);
-        createdUser.needsToAcceptTC = false;
+        _principalIds[principal] = user.id;
+        _createGuestUsers(ctx, principals, callback, _principalIds);
+    });
+};
+
+/**
+ * Verify a guest token and associate a local login strategy with the user matching the token.
```

Comment:
Maybe we can add a bit more explanation about where these tokens come from and how they are used